backends/graphs/controls.h

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef _BACKENDS_GRAPHS_CONTROLS_H_
+#define _BACKENDS_GRAPHS_CONTROLS_H_
+
+#include <boost/graph/graph_traits.hpp>
+#include <boost/graph/adjacency_list.hpp>
+
+#include <boost/optional.hpp>
+
+#include <string>
+#include <utility>  // std::pair
+#include <vector>
+
+#include "ir/ir.h"
+#include "ir/visitor.h"
+
+namespace graphs {
+
+class EdgeTypeIface {
+ public:
+    virtual ~EdgeTypeIface() { }
+    virtual std::string label() const = 0;
+};
+
+class NameStack {
+ public:
+    NameStack() = default;
+    void push_back(const cstring &name) {
+        stack.push_back(name);
+    }
+    void pop_back() {
+        stack.pop_back();
+    }
+    std::string get(const cstring &name) const {
+        std::stringstream sstream;
+        for (auto &n : stack) sstream << n << ".";
+        sstream << name;
+        return sstream.str();
+    }
+ private:
+    std::vector<cstring> stack;
+};
+
+class ControlGraphs : public Inspector {
+ public:
+    using edgeProperty = boost::property<boost::edge_name_t, std::string>;
+    using vertexProperty = boost::property<boost::vertex_name_t, std::string>;
+    using Graph = boost::adjacency_list<boost::vecS, boost::vecS,
+        boost::directedS, vertexProperty, edgeProperty>;
+    using vertex_t = boost::graph_traits<Graph>::vertex_descriptor;
+
+    using Parents = std::vector<std::pair<vertex_t, EdgeTypeIface *> >;
+
+    ControlGraphs(P4::ReferenceMap *refMap, P4::TypeMap *typeMap, const cstring &graphsDir);
+
+    // merge misc control statements (action calls, extern method calls,
+    // assignments) into a single vertex to reduce graph complexity
+    boost::optional<vertex_t> merge_other_statements_into_vertex();
+
+    vertex_t add_vertex(const std::string &name);
+    vertex_t add_vertex(const cstring &name);
+
+    bool preorder(const IR::PackageBlock *block) override;
+
```

Comment:
Heh, actually, after seeing the code *below* the portion that Mihai quoted (as opposed to the portion above, which is what shows up in the Github summary) I actually agree with Mihai on this. You're probably better off not having the blank line between every `preorder` method.

Sorry for the noise, Mihai. =)